File: backend/crud/equipment.py
from __future__ import annotations
from typing import List, Optional

from sqlalchemy.orm import Session
from sqlalchemy.exc  import SQLAlchemyError, IntegrityError
from sqlalchemy      import func

from backend.models.equipment   import Equipment
from backend.schemas.equipment  import EquipmentIn

# ───────────────────────────────────────────────
# READ
# ───────────────────────────────────────────────
def get_equipment_by_machine(
    db: Session,
    machine_id: str,
    site: str | None = None,          # ★ site 파라미터
) -> Optional[Equipment]:
    """
    machine_id 로 1행 조회
    • site 가 주어지면 동일 site 행만 찾는다.
      (대소문자 무시가 필요하면 func.lower 적용)
    """
    query = db.query(Equipment).filter(Equipment.machine_id == machine_id)
    if site:
        query = query.filter(Equipment.site == site)
        # 대소문자 무시는 ↓ 이런 식으로
        # query = query.filter(func.lower(Equipment.site) == site.lower())

    return query.first()


def get_all_equipment(db: Session, site: str | None = None) -> List[Equipment]:
    query = db.query(Equipment)
    if site:
        query = query.filter(Equipment.site == site)
    return query.order_by(Equipment.no.asc()).all()


# ───────────────────────────────────────────────
# CREATE / UPDATE / UPSERT
# ───────────────────────────────────────────────
def upsert_equipment(db: Session, data: EquipmentIn) -> Equipment:
    """
    machine_id + site 기준 UPSERT
    • 존재 → UPDATE
    • 미존재 → INSERT
    """
    equip = get_equipment_by_machine(db, data.machine_id, data.site)

    try:
        if equip is None:  # INSERT
            equip = Equipment(
                machine_id = data.machine_id,
                slot_code  = data.slot_code,
                site       = data.site,         # ★ INSERT 시 site 고정
            )
            db.add(equip)

        # 공통 필드 업데이트 (INSERT / UPDATE 공통)
        equip.progress       = data.progress
        equip.shipping_date  = data.shipping_date
        equip.customer       = data.customer
        equip.manager        = data.manager
        equip.note           = data.note
        equip.slot_code      = data.slot_code
        equip.site           = data.site        # ★ UPDATE 시에도 site 반영

        db.commit()
        db.refresh(equip)
        return equip

    except IntegrityError:
        db.rollback()
        raise
    except SQLAlchemyError:
        db.rollback()
        raise


# ───────────────────────────────────────────────
# DELETE (옵션)
# ───────────────────────────────────────────────
def delete_equipment(db: Session, machine_id: str, site: str | None = None) -> None:
    equip = get_equipment_by_machine(db, machine_id, site)
    if equip:
        db.delete(equip)
        db.commit()

File: frontend/src/Equipment/equipment.ts
/* ------------------------------------------------------------
 * Equipment API (/api/equipment) – FastAPI 백엔드와 통신
 * ---------------------------------------------------------- */
import { http } from '../lib/http';

/* ---------- 공용 DTO 타입 ---------- */
export interface EquipmentDTO {
  machineId:    string;
  progress:     number;
  shippingDate: string;  // YYYY-MM-DD
  customer:     string;
  manager:      string;
  note:         string;
  slotCode:     string;
  site:         string;  // '본사' | '부항리' | '진우리'
}

/* ----------------------------- GET -----------------------------
 * • machineId 로 단일 장비 조회
 * • site 쿼리스트링을 함께 보내 필터링 (예: /equipment/J-01-03?site=본사)
 * -------------------------------------------------------------- */
export const fetchEquipment = (machineId: string, site: string) =>
  http
    .get<EquipmentDTO>(`/equipment/${machineId}`, { params: { site } })
    .then(res => res.data);

/* ---------------------------- POST -----------------------------
 * • UPSERT: 존재하면 UPDATE, 없으면 INSERT
 * • payload 안에 site 필드가 포함돼 있으므로 경로 수정 불필요
 * -------------------------------------------------------------- */
export const saveEquipment = (data: EquipmentDTO) =>
  http.post<EquipmentDTO>('/equipment/', data).then(res => res.data);
